app/integration/runtime/src/main/java/io/syndesis/integration/runtime/handlers/TemplateStepHandler.java

```diff
@@ -0,0 +1,225 @@
+
+    private static final String DOUBLE_CLOSE_BRACE_PATTERN = "\\}\\}";
+
+    private static final String MUSTACHE_OPEN_DELIMITER = "[[";
+
+    private static final String MUSTACHE_CLOSE_DELIMITER = "]]";
+
+    private final JsonToMapProcessor jsonToMapProcessor = new JsonToMapProcessor();
+
+    private final TextToJsonProcessor textToJsonProcessor = new TextToJsonProcessor();
+
+    @Override
+    public boolean canHandle(Step step) {
+        if (StepKind.template != step.getStepKind()) {
+            return false;
+        }
+
+        Action action = step.getAction().orElse(null);
+        if (action == null) {
+            return false;
+        }
+
+        DataShape inputDataShape = action.getInputDataShape().orElse(null);
+        if (inputDataShape == null) {
+            return false;
+        }
+
+        return DataShapeKinds.JSON_SCHEMA == inputDataShape.getKind();
+    }
+
+    @SuppressWarnings("PMD.AvoidReassigningParameters")
+    @Override
+    public Optional<ProcessorDefinition<?>> handle(Step step, ProcessorDefinition<?> route, IntegrationRouteBuilder builder, String flowIndex, String stepIndex) {
+        ObjectHelper.notNull(route, "route");
+
+        Map<String, String> properties = step.getConfiguredProperties();
+        String template = properties.get(TEMPLATE_PROPERTY);
+
+        /*
+         * To avoid creating a temporary file, we use the mustache template header
+         * to override the camel-mustache resource uri. This is more efficient but has
+         * a downside.
+         *
+         * Once returned, the route (ProcessorDefinition) is resolved. This resolution
+         * looks for any properties delimited by {{ and }}, which is the default syntax
+         * for mustache properties. Thus, resolution of the mustache properties is
+         * incorrectly attempted and unsurprisingly fails since these properties are
+         * meant for mustache and not camel.
+         *
+         * Changing the mustache delimiter patterns avoids this problem and allows the
+         * properties to be resolved later by mustache.
+         */
+        template = template.replaceAll(DOUBLE_OPEN_BRACE_PATTERN, MUSTACHE_OPEN_DELIMITER);
+        template = template.replaceAll(DOUBLE_CLOSE_BRACE_PATTERN, MUSTACHE_CLOSE_DELIMITER);
+
+        //
+        // Convert the exchange's in message from JSON
+        // to a HashMap since this is required by camel modules
+        //
+        route = route.process(jsonToMapProcessor);
+
+        //
+        // Apply the template to the mustache header property
+        // Then add to the route path
+        //
+        route.setHeader(MustacheConstants.MUSTACHE_TEMPLATE).constant(template);
+
+        //
+        // Encode the delimiters since they are applied as URI query parameters
+        //
+        try {
+            String id = step.getId().orElse(UUID.randomUUID().toString()) + UNDERSCORE + UNDERSCORE + stepIndex;
```

Comment:
better not to use the step id as it may contain not allowed chars, a simple strategy would be to use something like:

    step-${flowIndex}-${stepIndex}

Reply:
done.